### PR TITLE
[UNI-125] Remove USBCDC/HWCDC defines

### PR DESCRIPTION
**Pull request by javijurado**
Remove defines for existing ones and switch from python to python3

### CODE REVIEW

File: rosserial_arduino/src/ros_lib/ArduinoHardware.h [hunk 1]
```diff
@@ -58,16 +58,17 @@
...
-  #else
+
+  #if ARDUINO_USB_CDC_ON_BOOT 
+    #if ARDUINO_USB_MODE 
+      #define SERIAL_CLASS HWCDC
+    #else
+      #define SERIAL_CLASS USBCDC
+    #endif  
+  #else  
+    #include <HardwareSerial.h>  // Arduino AVR
```
Comment (meyerj) [inline rosserial_arduino/src/ros_lib/ArduinoHardware.h]: I think you can include `HardwareSerial.h` unconditionally? `Serial` is an instance of the correct type then, as defined in
https://github.com/espressif/arduino-esp32/blob/cf448906b3836fbe9368934713b697469254c62f/cores/esp32/HardwareSerial.h#L346-L366

And that is also what gets assigned to `SERIAL_CLASS *iostream` in
https://github.com/Intermodalics/rosserial/blob/112cebcbd13076c90bb5d2506044a64d7f412724/rosserial_arduino/src/ros_lib/ArduinoHardware.h#L88
and hence has to work either way, unless you would add the preprocessor conditions there, too.

So
```cpp
#define SERIAL_CLASS decltype(Serial)
```
or
```cpp
typedef decltype(Serial) SERIAL_CLASS;
```
should work...

File: rosserial_arduino/src/ros_lib/ArduinoHardware.h [hunk 2]
```diff
@@ -58,16 +58,17 @@
...
     #define SERIAL_CLASS HardwareSerial
-  #endif
+  #endif
```
Comment (meyerj) [inline rosserial_arduino/src/ros_lib/ArduinoHardware.h]: nit: trailing space added
```suggestion
  #endif
```

**Review by meyerj — COMMENTED** [review]
One possible issue is that if `ARDUINO_USB_CDC_ON_BOOT` is defined (the default), then the `USBCDC` instance that `Serial` points to, defined [here](https://github.com/espressif/arduino-esp32/blob/cf448906b3836fbe9368934713b697469254c62f/cores/esp32/USBCDC.cpp#L435), was already initialized before. No need to call `iostream->init()` then anymore, or that may even be the reason for observed instabilities. So also patch
https://github.com/Intermodalics/rosserial/blob/112cebcbd13076c90bb5d2506044a64d7f412724/rosserial_arduino/src/ros_lib/ArduinoHardware.h#L107-L113
and not call `init()` there again if `ARDUINO_USB_CDC_ON_BOOT` is defined?

File: rosserial_arduino/src/ros_lib/ArduinoHardware.h (hunk 1 above)
Comment (javijurado) [reply rosserial_arduino/src/ros_lib/ArduinoHardware.h]: Thanks I will add HardwareSerial again on top of the ifdef.

> One possible issue is that if ARDUINO_USB_CDC_ON_BOOT is defined (the default), then the USBCDC instance that Serial points to, defined [here](https://github.com/espressif/arduino-esp32/blob/cf448906b3836fbe9368934713b697469254c62f/cores/esp32/USBCDC.cpp#L435), was already initialized before. No need to call iostream->init() then anymore, or that may even be the reason for observed instabilities. So also patch<

I can test it but this is the way that we have to initialize the baud rate, right?
But as you mention if ARDUINO_USB_CDC_ON_BOOT, we can omit this init because also on the HWCDC is initialized [here](https://github.com/espressif/arduino-esp32/blob/cf448906b3836fbe9368934713b697469254c62f/cores/esp32/HWCDC.cpp#L533)

Comment (javijurado) [reply rosserial_arduino/src/ros_lib/ArduinoHardware.h]: Tested it and it is still the same, it froze. The good news is that we can use `HWCDC` and also flash it using esptool libraries

Comment (meyerj) [reply rosserial_arduino/src/ros_lib/ArduinoHardware.h]: > I can test it but this is the way that we have to initialize the baud rate, right?

There is no baud rate in [CDC mode](https://en.wikipedia.org/wiki/USB_communications_device_class), either one. The argument in some calls is only for compatibility with `Serial.begin(baudrate)` calls in Arduino code if `Serial` would refer to a CDC device. Same on the Linux side.

File: rosserial_arduino/src/ros_lib/ArduinoHardware.h [hunk 3]
```diff
@@ -104,11 +104,11 @@ class ArduinoHardware {
...
-#if defined(USE_USBCON)
-      // Startup delay as a fail-safe to upload a new sketch
-      delay(3000); 
-#endif
-      iostream->begin(baud_);
+      #if defined(USE_USBCON)
+            // Startup delay as a fail-safe to upload a new sketch
+            delay(3000); 
+      #endif
+            iostream->begin(baud_);
```
Comment (meyerj) [inline rosserial_arduino/src/ros_lib/ArduinoHardware.h]: nit: unrelated whitespace-only change?

File: rosserial_arduino/src/ros_lib/ArduinoHardware.h [hunk 4]
```diff
@@ -57,14 +57,14 @@
...
-    // ESP32-S3
-    #define SERIAL_CLASS HWCDC
+#else
+  #if ARDUINO_USB_CDC_ON_BOOT 
+    #include <HardwareSerial.h>  // Arduino AVR
+    #if ARDUINO_USB_MODE 
+      #define SERIAL_CLASS HWCDC
+    #else
+      #define SERIAL_CLASS USBCDC
+    #endif
```
Comment (meyerj) [inline rosserial_arduino/src/ros_lib/ArduinoHardware.h]:
```suggestion
  #include <HardwareSerial.h>  // Arduino AVR
  #define SERIAL_CLASS decltype(Serial)
```

should work in any combination of preprocessor flags `ARDUINO_USB_CDC_ON_BOOT` and `ARDUINO_USB_MODE` without a change in behavior. The compiled code would be exactly the same. Just without the need to replicate the conditions in [`HardwareSerial.h`](https://github.com/espressif/arduino-esp32/blob/cf448906b3836fbe9368934713b697469254c62f/cores/esp32/HardwareSerial.h#L346-L353).

Calling `iostream->begin(baud_)` in `init()` or not is a different story. From the code in [`HWCDC::begin()`](https://github.com/espressif/arduino-esp32/blob/cf448906b3836fbe9368934713b697469254c62f/cores/esp32/HWCDC.cpp#L244-L246) I would still think that the USB host has to enumerate the device twice on reboot (disconnect - connect - disconnect - connect cycle), if `begin()` was called twice, once from `arduino-esp32`'s main function and once in `ArduinoHardware::init()`. But usually the ESP32 controller is already booted long before the PC and it _should_ never reboot, so it does not matter much.

**Review by meyerj — APPROVED** [review]

File: rosserial_arduino/src/ros_lib/ArduinoHardware.h [hunk 5]
```diff
@@ -104,11 +96,18 @@ class ArduinoHardware {
     int getBaud(){return baud_;}
 
     void init(){
-#if defined(USE_USBCON)
-      // Startup delay as a fail-safe to upload a new sketch
-      delay(3000); 
-#endif
-      iostream->begin(baud_);
+      #if defined(USE_USBCON)
+<<<<<<< HEAD
```
Comment (meyerj) [inline rosserial_arduino/src/ros_lib/ArduinoHardware.h]: Left conflict resolution markers?
